keps/sig-api-machinery/5000-api-linting-crd-schema-tooling/README.md

```diff
@@ -0,0 +1,522 @@
+
+### Goals
+
+- Create tooling to help developers writing Kubernetes APIs in Go to follow best practices and Kubernetes API conventions
+- Create tooling to help developers writing CustomResourceDefinitions avoid making breaking changes
+
+### Non-Goals
+
+- Create tooling for validating CustomResourceDefinition schema changes at admission time
+    - This is a non-goal of this KEP in particular to keep the scope of this KEP to only include the individual tooling. A separate KEP may be written, focusing on addressing this goal.
+
+## Proposal
+
+This KEP introduces two new sig-api-machinery subprojects.
+The projects are intended to complement each other, and take lessons learned from [Kubernetes API conventions][kube-api-conventions] and our experience
+with [making changes to Kubernetes APIs][making-changes-to-apis], to enable the wider Kubernetes ecosystem to follow best practices and avoid breaking changes.
+
+### Go-based Kubernetes API Linting
+
+A linter, built as a `golangci-lint` plugin, that evaluates Go type defintions (typically `_types.go`), and flags deviations from Kubernetes API conventions and best practices.
+
+The linter will be configurable, allowing users to choose which of the rules they wish to adhere to, and which they wish to ignore.
+
+The linter will focus, as much as possible on providing a "built-ins" experience by default, targetting rules for types in Kubernetes/Kubernetes.
+It will also provide configuration to allow for a more "CRD" experience, adjusting the configuration and rules to cater to the needs of CRD authors.
+
+The linter will focus on static evaluation of CRD types.
+It cannot, and will not detect changes to the code, and therefore cannot assert any rule that requires information about a previous version of the types.
+
+### CRD Schema Change Validation
+
+Create a CLI tool that evaluates the differences between an old and new CRD in YAML to identify changes that may break users.
+
+All validations are exported in Go packages that can be imported and consumed by other projects.
+
+Some examples of how the CLI may be used:
+
+- Comparing CRD that exists on cluster to a local YAML file of the same CRD - `crdify kube://mycrd.example.org file://mycrd.yaml`
+- Comparing CRD across git revisions (could be used in CI systems) - `crdify git://main?path=mycrd.yaml git://HEAD?path=mycrd.yaml`
+- Comparing CRD from git revision to a local file - `crdify git://main?path=mycrd.yaml file://mycrd.yaml`
+
+The schema change validation will focus on changes that may break users, and changes that can be detected by comparing the CRD schemas in YAML.
+It will not focus on static analysis, which will be the focus of the linter.
+
+### Why do we need both?
+
+When authoring types for Kubernetes, we typically use Go types, and add "markers" (e.g. `+optional`) to provide information about how the fields should behave.
```

Comment:
Just a heads up that https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/5073-declarative-validation-with-validation-gen/README.md is being implemented and aims to significantly increase the usage of markers for validation of built-in types.

I would expect the ideas being proposed in this KEP to have broader applications as more declarative validation tags are added to built-in types, since the amount of machine accessible information about built-in types will increase.

Reply:
I hadn't seen the proposal from that KEP before, I'll give that a read as this is exactly the kind of thing I think KAL will want to be supporting in the future.